build_runner/lib/src/asset_graph/graph.dart

```diff
@@ -94,38 +100,79 @@ class AssetGraph {
   /// set of assets that need to be deleted.
   Iterable<AssetId> updateAndInvalidate(Map<AssetId, ChangeType> updates) {
     var deletes = new Set<AssetId>();
-    var seen = new Set<AssetId>();
     void clearNodeAndDeps(AssetId id, ChangeType rootChangeType,
-        {AssetId parent}) {
-      if (seen.contains(id)) return;
-      seen.add(id);
+        {AssetId parent, bool rootIsSource}) {
       var node = this.get(id);
       if (node == null) return;
+      if (parent == null) rootIsSource = node is! GeneratedAssetNode;
 
       // Update all outputs of this asset as well.
       for (var output in node.outputs) {
-        clearNodeAndDeps(output, rootChangeType, parent: node.id);
+        clearNodeAndDeps(output, rootChangeType,
+            parent: node.id, rootIsSource: rootIsSource);
       }
 
-      if (parent == null && rootChangeType == ChangeType.REMOVE) {
-        // This is the root asset, it was removed on disk so it needs to be
-        // removed from the graph.
-        remove(id);
-      }
       if (node is GeneratedAssetNode) {
-        node.needsUpdate = true;
-        if (rootChangeType == ChangeType.REMOVE &&
+        if (rootIsSource &&
+            rootChangeType == ChangeType.REMOVE &&
             node.primaryInput == parent) {
-          remove(id);
+          _remove(id);
           deletes.add(id);
+        } else {
+          node.needsUpdate = true;
         }
+      } else {
+        // This is a source
+        if (rootChangeType == ChangeType.REMOVE) _remove(id);
       }
     }
 
     updates.forEach(clearNodeAndDeps);
     return deletes;
   }
 
+  /// Updates the structure of the graph to account for any new sources and
+  /// returns true if the graph has changed.
+  bool updateForSources(
+      List<List<BuildAction>> phases, Iterable<AssetId> newSources) {
+    var unseen = newSources.where((s) => !_nodesById.containsKey(s)).toList();
+    if (unseen.isEmpty) return false;
+    _addOutputsForSources(phases, unseen.toList());
+    return true;
+  }
+
+  void _addOutputsForSources(
+      List<List<BuildAction>> phases, Iterable<AssetId> newSources) {
+    newSources.map((s) => new AssetNode(s)).forEach(_add);
+    var allInputs = new Set<AssetId>()..addAll(newSources);
+    var phaseNumber = 0;
+    for (var phase in phases) {
+      phaseNumber++;
+      var phaseOutputs = <AssetId>[];
+      for (var action in phase) {
+        var inputs = allInputs.where(
+            (input) => action.inputSet.globs.any((g) => g.matches(input.path)));
+        for (var input in inputs) {
+          var outputs = expectedOutputs(action.builder, input);
+          phaseOutputs.addAll(outputs);
+          get(input).primaryOutputs.addAll(outputs);
+          for (var output in outputs) {
+            if (contains(output) && get(output) is AssetNode) {
+              _remove(output);
```

Comment:
Yes - if we see a generated file on disk in `build_impl.dart` we don't know that it was generated so we treat it as a source. Here when we see that it's actually generated we overwrite it as a generated node and expect that build_impl will clean it up as appropriate. We're mainly concerned here with getting the structure of the graph right, not with the state of the build or previous builds.